### PR TITLE
[INVE-17817] Move nodes await to make room for the expansion

### PR DESCRIPTION
**Pull request by Blakko**
With this PR the expansion function will automatically move the origin points away to make room for the expansion.
Additionally, this adds the possibility to customize the animation params for the `positionsAnimate` function

How to test:
- Download this demo app -> [demoreactapp-main.zip](https://github.com/sirensolutions/G6/files/7926327/demoreactapp-main.zip)
- run `npm install`
- Check out this branch, go into the packages/core folder and run `npm link`
- go to the demo app folder and run `npm link @antv/g6-core`
- now start the demo app with npm start

You'll end up with a single node on the screen. Pressing the `E` key will cause the expansion (some nodes should pop up on the top left corner). Now, pressing the `U` key will produce a console log showing the undo/redo stacks. You should see the expansion operation in the stack

Notes:
- You can change the number of initial nodes in thee `App.js` file -> `const sourceCount = 3;`
- You can change the number of expanded nodes per node in the `expansion.tsx` file -> `const NODE_COUNT = 30;`

https://user-images.githubusercontent.com/2861371/151809884-de2dd952-1243-4789-888e-b72c3175d8e8.mp4

### CODE REVIEW

File: packages/core/src/graph/graph.ts [hunk 1]
```diff
@@ -1426,6 +1479,8 @@ export default abstract class AbstractGraph extends EventEmitter implements IAbs
       });
     }
 
+    this.positionsAnimate(animateCfg);
```
Comment (niknbr) [inline packages/core/src/graph/graph.ts:1482]: I guess we have to check if animation flag is on here ? Else resort to [refreshPositions()](https://g6.antv.vision/en/docs/api/graphFunc/item#graphrefreshpositions)

File: packages/core/src/graph/graph.ts [hunk 2]
```diff
@@ -1350,7 +1351,60 @@ export default abstract class AbstractGraph extends EventEmitter implements IAbs
...
+      }
+      return { x: x1 * (1.0 - factor) + x2 * factor, y: y1 * (1.0 - factor) + y2 * factor };
+    }
+    // Move the entry away from all the previously processed origins
+    function movedEntryIfNeeded(entry, prevOrigins) {
+      const sourceModel = entry.model;
+      let moved = false;
+      // compute distance with previously processed nodes
+      for (var p = 0; p < prevOrigins.length; p++) {
+        const procEntry = prevOrigins[p]
```
Comment (niknbr) [inline packages/core/src/graph/graph.ts:1370]: `;`

File: packages/core/src/graph/graph.ts [hunk 3]
```diff
@@ -1350,7 +1351,60 @@ export default abstract class AbstractGraph extends EventEmitter implements IAbs
...
+        const procModel = procEntry.model;
+        const minDistance = entry.maxRadius + procEntry.maxRadius + levelOffset / 2;
+
+        //If the x/y are the same offset them a little to avoid infinite loops
+        sourceModel.x = sourceModel.x === procModel.x ? sourceModel.x + 10 : sourceModel.x;
+        sourceModel.y = sourceModel.y === procModel.y ? sourceModel.y + 10 : sourceModel.y;
+        const a = sourceModel.x - procModel.x;
+        const b = sourceModel.y - procModel.y;
+
+        let distance = Math.abs(Math.sqrt(a * a + b * b));
```
Comment (niknbr) [inline packages/core/src/graph/graph.ts:1380]: Suggestion 🙂 
```javascript
const distance = Math.max(Math.abs(Math.sqrt(a * a + b * b)), 1)
```

File: packages/core/src/graph/graph.ts (hunk 1 above)
Comment (Blakko) [reply packages/core/src/graph/graph.ts:1482]: the animate flag is always external to the GraphAnimateConfig, but I agree that for consistency with the other functions it should be possible to disable the animation

**Review by niknbr — APPROVED** [review]
LGTM

**Pull request merged**